### PR TITLE
Reconnect TCP connections on failure

### PR DESCRIPTION
**Pull request by LasseGosch**

### CODE REVIEW

**Comment by lars-sh** [comment]
I'd expect to use no retries, but resetting the writer and rethrowing the original IOException. That even simplifies the code a lot and does not require sleeps. The outer code can handle the error accordingly. Example:

```
} catch (IOException) {
    close();
    throw e;
}
```

And inside the close method you could reset the socket and writer fields to `Optional.empty()`.

File: log4jSyslogWriter64k/src/main/java/com/github/loggly/log4j/helpers/SyslogTcpWriter64k.java [hunk 1]
```diff
@@ -70,5 +76,8 @@ public void close() throws IOException {
 		if (socket.isPresent()) {
 			socket.get().close();
 		}
+
+		writer = Optional.empty();
+		socket = Optional.empty();
```
Comment (lars-sh) [inline log4jSyslogWriter64k/src/main/java/com/github/loggly/log4j/helpers/SyslogTcpWriter64k.java:81]: Please do this in a finally block as there might occur additional IOExceptions in error cases.

File: log4jSyslogWriter64k/src/main/java/com/github/loggly/log4j/helpers/SyslogTcpWriter64k.java [hunk 2]
```diff
@@ -45,8 +47,12 @@ public void write(final char[] buf, final int off, final int len) throws IOExcep
 	public void write(final String string) throws IOException {
 		// compute syslog frame according to: https://tools.ietf.org/html/rfc6587
 		final String syslogFrame = String.format("%s %s", string.length(), string);
-
-		getWriter().append(syslogFrame);
+		try {
+			getWriter().append(syslogFrame);
+		} catch (final IOException e) {
+			close();
```
Comment (lars-sh) [inline log4jSyslogWriter64k/src/main/java/com/github/loggly/log4j/helpers/SyslogTcpWriter64k.java:53]: This should probably be wrapped in a try-catch, ignoring IOExceptions coming from the close method to make sure, exceptions inside the close method do not hide the original IOException from here.

**Review by lars-sh — CHANGES_REQUESTED** [review]

**Review by lars-sh — APPROVED** [review]

**Comment by lars-sh** [comment]
@noxone, please review and approve this PR, esp. the synchronized handling. Thanks :)

**Review by noxone — APPROVED** [review]

**Pull request merged**